src/XR/features/WebXRFeaturePointSystem.ts

```diff
@@ -0,0 +1,176 @@
+     */
+    public readonly onFeaturePointsUpdatedObservable: Observable<number[]> = new Observable();
+    /**
+     * The currrent feature point cloud maintained across frames.
+     */
+    public readonly featurePointCloud: Array<IWebXRFeaturePoint> = this._featurePointCloud;
+
+    /**
+     * construct the feature point system
+     * @param _xrSessionManager an instance of xr Session manager
+     */
+    constructor(_xrSessionManager: WebXRSessionManager) {
+        super(_xrSessionManager);
+        if (this._xrSessionManager.session) {
+            this._init();
+        } else {
+            this._xrSessionManager.onXRSessionInit.addOnce(() => {
+                this._init();
+            });
+        }
+    }
+
+    /**
+     * Detach this feature.
+     * Will usually be called by the features manager
+     *
+     * @returns true if successful.
+     */
+    public detach(): boolean {
+        if (!super.detach()) {
+            return false;
+        }
+
+        this.featurePointCloud.length = 0;
+        return true;
+    }
+
+    /**
+     * Dispose this feature and all of the resources attached
+     */
+    public dispose(): void {
+        super.dispose();
+
+        this._featurePointCloud.length = 0;
+        this.onFeaturePointsUpdatedObservable.clear();
+    }
+
+    /**
+     * On receiving a new XR frame if this feature is attached notify observers new feature point data is available.
+     */
+    protected _onXRFrame(frame: XRFrame) {
+        if (!this.attached || !this._enabled || !frame) {
+            return;
+        }
+
+        const featurePointRawData: number[] | undefined = frame.featurePointCloud;
+        if (!featurePointRawData || featurePointRawData.length == 0) {
+            return;
+        } else {
+            if (featurePointRawData.length % 5 != 0) {
+                throw new Error("Received malformed feature point cloud of length: " + featurePointRawData.length);
+            }
+
+            const numberOfFeaturePoints : number = featurePointRawData.length / 5;
+            let updatedFeaturePoints = new Array();
+            let addedFeaturePoints = new Array();
+            for (var i = 0; i < numberOfFeaturePoints; i++) {
+                const rawIndex: number = i * 5;
+
+                const id = featurePointRawData[rawIndex + 4];
+                // IDs should be durable across frames and strictly increasing from 0 up, so use them as indexing into the feature point array.
+                if (id == this._featurePointCloud.length) {
+                    this._featurePointCloud.push({
+                        position: new Vector3(
+                             featurePointRawData[rawIndex],
+                             featurePointRawData[rawIndex + 1],
+                             featurePointRawData[rawIndex + 2]),
+                        confidenceValue: featurePointRawData[rawIndex + 3],
+                        id: id

```

Comment:
If we are talking about hundred-thousands of those points, we should probably try to be very "cheap" on space used. is the ID here absolutely needed? A developer accessing it will surely have a way to get the ID in a different way.

Reply:
Makes sense, we can remove ID here since we are getting it by id reference in the first place.